File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


vector<pair<int, int>> vec;
int dp[1002];
int arr[1002];

int main()
{
    int n;
    cin >> n;
    for(int i = 0; i < n; i++) cin >> arr[i];
    
    for(int i = 0; i < n; i++){
        int mx = 0;
        for(int j = 0; j < i; j++){
            if(arr[j] >= arr[i]) continue;
            mx = max(dp[j], mx);
        }
        dp[i] = max(dp[i], mx + 1);
    }
    
    int mx = 0;
    int idx = -1;
    for(int i = 0; i < n; i++){
        if(dp[i] > mx){
            mx = dp[i];
            idx = i;
        }
    }

    int cur = mx;
    vector<int> path;
    for(int i = idx; i >= 0; i--){
        if(cur == dp[i]){
            path.push_back(arr[i]);
            cur--;
        }
    }
    
    reverse(path.begin(), path.end());
    cout << path.size() << "\n";
    for(int v: path) cout << v << " ";
    
    return 0;
}

File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();

		int arr[] = new int[1002];
		for (int i = 0; i < N; i++) {
			arr[i] = sc.nextInt();
		}

		int dp[] = new int[1002];
		int pre[] = new int[1002];

		for (int i = 0; i < N; i++) {
			int mx = 0;
			for (int j = 0; j < i; j++) {
				if (arr[i] > arr[j] && dp[j] > mx) {
					mx = dp[j];
					pre[i] = j;
				
				}
			}
			dp[i] = mx + 1;
		}

		int cnt = 0;
		int idx = -1;
		for (int i = 0; i < N; i++) {
			if(dp[i] > cnt) {
				cnt = dp[i];
				idx =i;
			}

		}
		List<Integer> path = new ArrayList<>();

		int cur = idx;
		
		for(int i = 0; i < cnt; i++) {
			path.add(arr[cur]);
			cur = pre[cur];
		}

		Collections.reverse(path);

		StringBuilder sb = new StringBuilder();
		sb.append(cnt).append("\n");
		for (int i : path) {
			sb.append(i).append(" ");
		}
		System.out.println(sb);
	}

}